### ISSUE TITLE
Menu unification and rally to push mainline.

### ISSUE THREAD

**Issue by monkeyiq** [fontforge/fontforge#2734]
This is to track an update to https://github.com/fontforge/fontforge/issues/2257 for a menu unification for the charview, fontview, and metricsview. 

Interested parties are invited to explore the code in my public branch at https://github.com/monkeyiq/fontforge-fork/tree/2016menus

From a developer perspective this code introduces the new sharedmenu.c which contains the unification of the menu code from all three files: charview.c, fontview.c and metricsview.c. There is a crude type system and a new vtable (contained in sharedmenu_funcs) to allow each type of window to service the menu calls as is best done by the window. Some calls, such as almost all of the entries in the "File" menu can be serviced by a single code path. That code gets around being called with a 'gw' containing the userdata pointer for different window types by using the new type system and calls from classtypeui.h such as tryObtainGDataFontView(). This call will convert a CharView\* to a FontView\* by returning the delegated fv pointer contained in the charview. The calling code doesn't need to care, it just gets a correct fontview pointer from all three window types and can operate from there. All cv, mv, and fv can be converted to a fontview because all either are a fontview or contain a pointer to one.

Window types that allow zero, one, or more characters to be selected can implement the getActiveSplineChar() virtual function to tell the caller if 1+ chars are selected. This way, menu code and other interested callers can check/find the active character without needing to care what window or class type they are dealing with.

The type system, object coercion, and virtual table functions setup the code base to allow single functions to be applied to multiple window types. At the moment, there are some menuitems in some windows which are disabled. Some of these do not really have to be disabled, the code behind that menu item could possibly be updated to work across more window types than it was originally created for. For example, something that used to only be offered on the charview window might be enabled in the menu for the fontview window. Selecting that item in the fontview window would simply apply the same function that the charview menu item does, but for all chars that are selected in the fontview window that the menu item was selected from. This case is exactly what the new visitor mechanism visitAllCharViews() in fontview.c has been created to assist with. Guidance is requested as to which menu items should be enabled in which window for what menu items first. Some things are likely to be more urgent than others.

I have gone through the menus on the three window types checking for crashes but might have missed some, especially when the menu item is contingent on previous setup. I have also tried to unify the code that disables the menu items depending on if they should be made available in that window (ie, the code supports their use and they made sense).
### TODO
- [ ] maybe enable some more menuitems which are currently disabled.
- [ ] update code for some menu items which no longer work for some window+menu combinations
- [ ] Update the hotkeys to work with the new menu layout
- [ ] context menus
- [ ] I notice the right click context menu in charview window doesn't show point types properly for spiro mode.

**Comment by davelab6** [comment]
This is really awesome! Thanks @monkeyiq :) 

cc @larsenwork @pathumego @frank-trampe

**Comment by davelab6** [comment]
@pathumego when you have some spare time, I hope you can test Ben's branch for crashes as he mentions. Please post here if you need help getting his branch to run :)

**Comment by pathumego** [comment]
@monkeyiq I'll test this over the weekend. Ill have to build from the branch on Mac.

**Comment by pathumego** [comment]
@davelab6 @monkeyiq Can you post how to get this running on Mac and Win? I have a few people to test.

**Comment by pathumego** [comment]
@monkeyiq I ran the normal build process as in https://github.com/fontforge/fontforge/blob/master/INSTALL-git.md on Mac. I ran into few issues.
 
I ran `./configure` with solution listed in Common problems because I got the following: 
```
No package 'pangoxft' found

Consider adjusting the PKG_CONFIG_PATH environment variable if you
installed software in a non-standard prefix.
```
After that I ran `make` and got a bunch of warnings and following errors.
```
29 warnings generated.
  CC       libfontforge_la-unicoderange.lo
  CC       libfontforge_la-utils.lo
  CC       libfontforge_la-winfonts.lo
  CC       libfontforge_la-woff.lo
  CC       libfontforge_la-glif_name_hash.lo
  CC       libfontforge_la-classtype.lo
  CCLD     libfontforge.la
clang: warning: argument unused during compilation: '-pthread'
clang: warning: argument unused during compilation: '-pthread'
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: for architecture: x86_64 file: .libs/libfontforge.a(libfontforge_la-bezctx_ff.o) has no symbols
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: for architecture: x86_64 file: .libs/libfontforge.a(libfontforge_la-utils.o) has no symbols
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: for architecture: x86_64 file: .libs/libfontforge.a(libfontforge_la-classtype.o) has no symbols
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: for architecture: x86_64 file: .libs/libfontforge.a(fd-hook.o) has no symbols
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: for architecture: x86_64 file: .libs/libfontforge.a(sys_socket.o) has no symbols
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: for architecture: x86_64 file: .libs/libfontforge.a(unistd.o) has no symbols
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: for architecture: x86_64 file: .libs/libfontforge.a(xsize.o) has no symbols
Making all in fontforgeexe
Making all in pixmaps
Making all in tango
make[4]: Nothing to be done for `all-am'.
  CC       libfontforgeexe_la-startnoui.lo
  CCLD     libfontforgeexe.la

*** Warning: Linking the shared library libfontforgeexe.la against the non-libtool
*** objects  macobjective.o is not portable!
clang: warning: argument unused during compilation: '-pthread'
clang: warning: argument unused during compilation: '-pthread'
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: .libs/libfontforgeexe.a(fd-hook.o) has no symbols
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: .libs/libfontforgeexe.a(sys_socket.o) has no symbols
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: .libs/libfontforgeexe.a(unistd.o) has no symbols
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: .libs/libfontforgeexe.a(xsize.o) has no symbols
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: .libs/libfontforgeexe.a(fd-hook.o) has no symbols
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: .libs/libfontforgeexe.a(sys_socket.o) has no symbols
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: .libs/libfontforgeexe.a(unistd.o) has no symbols
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: .libs/libfontforgeexe.a(xsize.o) has no symbols
make[3]: `.gitignore' is up to date.
Making all in pyhook
  CC       psMatpyhook.lo
  CCLD     psMat.la
clang: warning: argument unused during compilation: '-pthread'
  CC       fontforgepyhook.lo
  CCLD     fontforge.la
clang: warning: argument unused during compilation: '-pthread'
Making all in mackeys
msgfmt --check --output-file=en.mo en.po
make[2]: msgfmt: No such file or directory
make[2]: *** [en.mo] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2
fontforge-fork$ 

```

**Comment by davelab6** [comment]
You should use homebrew to do this, but edit the recipe to use Ben's repo
and branch.

$ brew edit fontforge

**Comment by pathumego** [comment]
@davelab6 But the fontforge homebrew formula only install the command line tools.

**Comment by davelab6** [comment]
On 15 July 2016 at 09:50, Pathum Egodawatta <notifications@github.com>
wrote:

> @davelab6 <https://github.com/davelab6> But the fontforge homebrew
> formula only install the command line tools.
>
Ah, that sucks. Sorry :)

**Comment by monkeyiq** [comment]
I'll take a look at doing a local build and upload. Probably more interesting for you to use FontForge and look for issues rather than fighting a build.

**Comment by davelab6** [comment]
I think it would be much better to document your local build so that Pathum
can reproduce the process. He may wish to edit your patch with minor
adjustments (like ordering of items) and will need to compile his own
builds to do so.

**Comment by monkeyiq** [comment]
cool, will do. I've still been using macports for my local build. Just because that is what I started using for the local build.

**Comment by monkeyiq** [comment]
I'll also dig around. IIRC the macports build process is/was already documented, so might only need a touch up.

**Comment by monkeyiq** [comment]
@pathumego ok, so first thing that is stopping the build you are doing is the missing msgfmt. It appears that the tool is installed 'key only' by default because there might be one from apple. Since you have none then there is no contention there.

You can either stuff the path to these msgfmt into your PATH env variable or tell homebrew to make links to these programs inside its bin directory using the below.

```
brew link gettext
```

you might like to check from time to time for multiple versions using 
```
type -all msgfmt 
```
which will tell you if/when you have any second versions of msgfmt installed too.

Next on my radar is working out why pango xft doesn't get seen by your build.

**Comment by jtanx** [comment]
@monkeyiq from what I remember, default X11 support was dropped in homebrew hence the removal of pangoxft support - see #2737, and more importantly #2380.

**Comment by monkeyiq** [comment]
ok, I knew I had some doco for the macports build already, that is at https://github.com/fontforge/fontforge/tree/master/osx

I'm now taking a look at that to see how stale it is, and if the Portfile needs a little touchup to work on El Cap and a mid 2016 environment. That should also give an X11 enabled FontForge on osx.

**Comment by pathumego** [comment]
@monkeyiq Ok I linked with `brew link gettext` The pangoxft problem is still there. But when I ran configure with 
```
PYTHON_CFLAGS="-I/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7" \
PYTHON_LIBS="-L/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/config" \
./configure \
--disable-programs \
--disable-native-scripting \
--without-x \
--without-cairo \
--without-giflib \
--without-libjpeg \
--without-libtiff \
--without-libpng \
--without-libspiro \
--without-libuninameslist \
--without-libunicodenames \
--without-iconv \
--without-libzmq \
--without-libreadline \
--enable-python-scripting \
--enable-python-extension;
```
As per the doc, it worked without errors, but still I get a bunch of warnings. Should I go foreward on this or wait for you to figure out the `pangoxft` problem?

**Comment by pathumego** [comment]
Ok on macports build with your fork following `https://github.com/fontforge/fontforge/tree/master/osx` gives me following, Maybe will help you to make it work :) 

```
osx$ sudo ./force-rebuild.sh
Password:
... updating local git repo from github ... 
Already up-to-date.
/usr/local/src/github-fontforge/fontforge/osx
--->  Cleaning fontforge
--->  Cleaning fontforge
... starting build ... 
--->  Computing dependencies for fontforge.
--->  Fetching archive for fontforge
--->  fontforge-2.0.0_beta1_1468929005.darwin_15.x86_64.tbz2 doesn't seem to exist in /opt/local/var/macports/incoming/verified
--->  Attempting to fetch fontforge-2.0.0_beta1_1468929005.darwin_15.x86_64.tbz2 from https://packages.macports.org/fontforge
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0--->  Attempting to fetch fontforge-2.0.0_beta1_1468929005.darwin_15.x86_64.tbz2 from http://mse.uk.packages.macports.org/sites/packages.macports.org/fontforge
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   305  100   305    0     0  11967      0 --:--:-- --:--:-- --:--:-- 12200
--->  Attempting to fetch fontforge-2.0.0_beta1_1468929005.darwin_15.x86_64.tbz2 from http://lil.fr.packages.macports.org/fontforge
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0--->  Fetching distfiles for fontforge
--->  Verifying checksums for fontforge
--->  Extracting fontforge
sh: ./bootstrap: No such file or directory
Command failed: cd /opt/local/var/macports/build/_usr_local_src_github-fontforge_fontforge_osx/fontforge/work/fontforge-2.0.0_beta1 && ./bootstrap 
Exit code: 127
Error: org.macports.extract for port fontforge returned: command execution failed
Warning: targets not executed for fontforge: org.macports.activate org.macports.extract org.macports.patch org.macports.configure org.macports.build org.macports.destroot org.macports.install
Please see the log file for port fontforge for details:
    /opt/local/var/macports/logs/_usr_local_src_github-fontforge_fontforge_osx/fontforge/main.log
To report a bug, follow the instructions in the guide:
    http://guide.macports.org/#project.tickets
Error: Processing of port fontforge failed
```

**Comment by pathumego** [comment]
@monkeyiq Ping :)

**Comment by skef** [comment]
Dead project

**Comment by davelab6** [comment]
Are you saying the current menu structure is ideal, and the menu structure
here was inferior?

I think the current menu structure is poor, and can be improved; however,
to help users who are familiar with the current structure and want to
control their migration to a better structure, perhaps postponing it
forever, it seems to me that making the structure runtime configurable is
desirable.

**Comment by skef** [comment]
@davelab6 The work described in various menu unification issues was completed, the PR was #3638. I did the last bits in April because the issues hadn't been updated with the additional requirements and I thought I was helping. 

I said it was "dead", rather than "done", because there are no current plans to implement those additional requirements. That could of course change. But either way the milestones listed in these bugs are no longer relevant because they have been implemented. I found places for the last few missing items and restructured the enabling code so that everything was self-consistent. And I agree that the result was much easier to understand and use than the current menus.

**Comment by skef** [comment]
I guess to clarify:

1. No, I don't think the current menu structure is ideal
2. For various reasons I won't go into (but mostly having to do with the enabling code) the 2016,2019 project and a project to allow menus to be configurable have very little overlap of implementation. A few bits and pieces could be moved forward from the #3638 branch, but that's it. It's a different project that will need its own set of issues.

**Comment by davelab6** [comment]
> I guess to clarify:
> 
> 1. No, I don't think the current menu structure is ideal

I'm curious if you or anyone here is interested in improving it further :)

> 2. For various reasons I won't go into (but mostly having to do with the enabling code) the 2016,2019 project and a project to allow menus to be configurable have very little overlap of implementation. 

Oh! I didn't appreciate this. Thanks for pointing this out

> A few bits and pieces could be moved forward from the #3638 branch, but that's it. It's a different project that will need its own set of issues.

Totally fair :)
